File: poh2/poh2.cs
using System;
using System.Linq;
class PaizaPOH2{
	const int SIZE=9999999;
	static byte[] z=new byte[SIZE];
	static int input_count=0;

	static int get(){
		int r=0;
		for(;'0'<=z[input_count]&&z[input_count]<='9';)r=r*10+z[input_count++]-'0';
		input_count++;
		return r;
	}
	static int getC(){
		return z[input_count++];
	}
	static void Main(){
		Console.OpenStandardInput().Read(z,0,SIZE);
		int h,w,s,t,i,j,k;
		//int[] arg=Console.ReadLine().Split(' ').Select(_=>int.Parse(_)).ToArray();
		//h=arg[0];
		//w=arg[1];
		h=get();
		w=get();
		int[,,] a=new int[h,w,h];
		int[,] accum=new int[h,w+1];
		for(i=0;i<h;i++){
			//string line=Console.ReadLine();
			for(j=0;j<w;j++)a[i,j,0]=(getC()-'0')^1;
			getC();
		}
		for(i=0;i<h;i++){
			accum[0,a[i,0,0]]++;
			for(j=1;j<w;j++){
				if(a[i,j,0]>0){
					a[i,j,0]+=a[i,j-1,0];
					accum[0,a[i,j,0]]++;
				}
			}
		}
		for(i=1;i<h;i++)for(j=0;j<w;j++)for(k=1;k<=i;k++){
			a[i,j,k]=Math.Min(a[i-1,j,k-1],a[i,j,k-1]);
			if(a[i,j,k]==0)break;
			accum[k,a[i,j,k]]++;
		}
		for(i=0;i<h;i++)for(j=w-1;j>=0;j--)accum[i,j]+=accum[i,j+1];
		//k=int.Parse(Console.ReadLine());
		k=get();
		for(;k>0;k--){
			//arg=Console.ReadLine().Split(' ').Select(_=>int.Parse(_)).ToArray();
			//s=arg[0];
			//t=arg[1];
			s=get();
			t=get();
			Console.WriteLine(s<=h&&t<=w?accum[s-1,t]:0);
		}
	}
}

File: poh2/poh2.vb
module PaizaPOH2
	dim SIZE as integer=9999999
	dim z(SIZE-1) as byte
	dim input_count as integer=0

	function myget() as integer
		dim r as integer
		while 48<=z(input_count) andalso z(input_count)<=57
			r=r*10+z(input_count)-48
			input_count+=1
		end while
		input_count+=1
		return r
	end function
	function mygetC() as integer
		input_count+=1
		return z(input_count-1)
	end function

	sub Main()
		Console.OpenStandardInput().Read(z,0,SIZE)
		dim h as integer=myget()
		dim w as integer=myget()
		dim a(h-1,w-1,h-1) as integer
		dim accum(h-1,w) as integer
		for i as integer=0 to h-1
			for j as integer=0 to w-1
				a(i,j,0)=(mygetC()-48) xor 1
			next
			mygetC()
		next
		for i as integer=0 to h-1
			accum(0,a(i,0,0))+=1
			for j as integer=1 to w-1
				if a(i,j,0)>0
					a(i,j,0)+=a(i,j-1,0)
					accum(0,a(i,j,0))+=1
				end if
			next
		next
		for i as integer=1 to h-1
			for j as integer=0 to w-1
				for k as integer=1 to i
					a(i,j,k)=Math.min(a(i-1,j,k-1),a(i,j,k-1))
					if a(i,j,k)=0
						exit for
					end if
					accum(k,a(i,j,k))+=1
				next
			next
		next
		for i as integer=0 to h-1
			for j as integer=w-1 to 0 step -1
				accum(i,j)+=accum(i,j+1)
			next
		next
		dim k as integer=myget()
		while k>0
			dim s as integer=myget()
			dim t as integer=myget()
			Console.WriteLine(if(s<=h andalso t<=w,accum(s-1,t),0))
			k-=1
		end while
	end sub
end module